### PR TITLE
This adds an option which lists the files emitted by the compiler

### PR DESCRIPTION
**Pull request by paulvanbrenk**
This adds an option which lists the files emitted by the compiler this is
usefull in various scenarios to make publishing the generated files easier.

This will reduce code duplicated in the MSBuild task used for publishing websites. e.g. to Azure.

### CODE REVIEW

**Comment by mhegazy** [comment]
We can not use this in MSBuild task. the main issue is finding the list of emitted files means we need to run the compiler, but we need the list to know if the input files and newer than output files, so that incremental build knows if it needs to run this task. it is a catch 22.

what we have now is an approximation, it is not guaranteed to work correctly (e.g. it does not capture modules vs script files). but it is a fast way to avoid running a full compilation if it is not needed. 

if we want to put this logic in the compiler we will need to implement the flag so that it does not parse/compile, it "estimates" the outputs based on inputs, which is more or less what we have in the task today.

so long story short, i do not think it is worth it to add this flag.

**Comment by paulvanbrenk** [comment]
We still need this for projects with a `tsconfig.json` file, since we don't have any logic for that in the current task.. I rather have the logic in 1 place, i.e. the compiler.

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -322,11 +322,22 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
         let newLine = host.getNewLine();
-        let jsxDesugaring = host.getCompilerOptions().jsx !== JsxEmit.Preserve;
+        let jsxDesugaring = compilerOptions.jsx !== JsxEmit.Preserve;
         let shouldEmitJsx = (s: SourceFile) => (s.languageVariant === LanguageVariant.JSX && !jsxDesugaring);
 
+        // wrap the writeFile function we have on the host,
+        // so we can easily keep track of all the files written
+        if (compilerOptions.listEmittedFiles) {
+            let writeFileCore = host.writeFile;
+            host.writeFile = (fileName: string, data: string, writeByteOrderMark: boolean, onError?: (message: string) => void) => {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I don't think you should override the host. I think you should call an alternative function instead of `host.writeFile`

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -322,11 +322,22 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
         let diagnostics: Diagnostic[] = [];
         let newLine = host.getNewLine();
-        let jsxDesugaring = host.getCompilerOptions().jsx !== JsxEmit.Preserve;
+        let jsxDesugaring = compilerOptions.jsx !== JsxEmit.Preserve;
         let shouldEmitJsx = (s: SourceFile) => (s.languageVariant === LanguageVariant.JSX && !jsxDesugaring);
 
+        // wrap the writeFile function we have on the host,
+        // so we can easily keep track of all the files written
+        if (compilerOptions.listEmittedFiles) {
+            let writeFileCore = host.writeFile;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This orphans `writeFile` from the `host`, so you end up losing context.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -545,7 +545,7 @@ namespace ts {
             // immediately bail out.  Note that we pass 'undefined' for 'sourceFile' so that we
             // get any preEmit diagnostics, not just the ones
             if (options.noEmitOnError && getPreEmitDiagnostics(program, /*sourceFile:*/ undefined, cancellationToken).length > 0) {
-                return { diagnostics: [], sourceMaps: undefined, emitSkipped: true };
+                return { diagnostics: [], sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Just make an empty array here.

File: src/compiler/tsc.ts [hunk 4]
```diff
@@ -105,15 +105,33 @@ namespace ts {
         }
     }
 
+    function reportEmittedFiles(fileNames: string[], host: CompilerHost) {
+        for (let i = 0; i < fileNames.length; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Use a for-of.

File: src/compiler/tsc.ts [hunk 5]
```diff
@@ -105,15 +105,33 @@ namespace ts {
...
 
+    function reportEmittedFiles(fileNames: string[], host: CompilerHost) {
+        for (let i = 0; i < fileNames.length; i++) {
+            reportEmittedFile(fileNames[i], host);
+        }
+    }
+
+    function reportEmittedFile(fileName: string, host: CompilerHost): void {
+
+        let category = DiagnosticCategory[DiagnosticCategory.Message].toLowerCase();
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Use `const`

File: src/compiler/tsc.ts [hunk 6]
```diff
@@ -105,15 +105,33 @@ namespace ts {
...
+        const relativeFileName = host
+            ? convertToRelativePath(fileName, host.getCurrentDirectory(), fn => host.getCanonicalFileName(fn))
+            : fileName;
+
+        let output = `${category} TSFILE: ${relativeFileName}${sys.newLine}`;
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: `const`

File: src/compiler/tsc.ts [hunk 7]
```diff
@@ -105,15 +105,33 @@ namespace ts {
...
     }
 
+    function reportEmittedFiles(fileNames: string[], host: CompilerHost) {
+        for (let i = 0; i < fileNames.length; i++) {
+            reportEmittedFile(fileNames[i], host);
+        }
+    }
+
+    function reportEmittedFile(fileName: string, host: CompilerHost): void {
+
```
Comment (DanielRosenwasser) [inline src/compiler/tsc.ts]: Remove line

**Comment by DanielRosenwasser** [comment]
:-1:

**Comment by mhegazy** [comment]
I do not think we should have two ways of computing output files, why is packaging not working?

**Comment by paulvanbrenk** [comment]
Talked to Mohamed and looks like we'll be taking a different approach.. closing this pull request for now.
